asllib/Typing.ml

```diff
@@ -578,42 +592,44 @@ module Annotate (C : ANNOTATE_CONFIG) = struct
                "is" *)
-            let struct1 = Types.get_structure env t1
-            and struct2 = Types.get_structure env t2 in
+            let struct1 =
+              Types.get_structure env t1 |> Types.to_well_constrained
+            and struct2 =
+              Types.get_structure env t2 |> Types.to_well_constrained
+            in
             match (struct1.desc, struct2.desc) with
-            | T_Int None, T_Int _ | T_Int _, T_Int None ->
+            | T_Int UnConstrained, T_Int _ | T_Int _, T_Int UnConstrained ->
                 (* Rule ZYWY: If both operands of an integer binary primitive
                    operator are integers and at least one of them is an
                    unconstrained integer then the result shall be an
                    unconstrained integer. *)
                 (* TODO: check that no other checks are necessary. *)
-                T_Int None |> with_loc
-            | T_Int (Some []), T_Int (Some _) | T_Int (Some _), T_Int (Some [])
-              ->
-                (* Rule BZKW: If both operands of an integer binary primitive
-                   operator are constrained integers and at least one of them
-                   is the under-constrained integer then the result shall be an
-                   under-constrained integer. *)
-                T_Int (Some []) |> with_loc
-            | T_Int (Some cs1), T_Int (Some cs2) ->
+                T_Int UnConstrained |> with_loc
+            | T_Int (UnderConstrained _), _ | _, T_Int (UnderConstrained _) ->
+                assert false (* We used to_well_constrained before *)
+            | T_Int (WellConstrained cs1), T_Int (WellConstrained cs2) ->
                 (* Rule KFYS: If both operands of an integer binary primitive
                    operation are well-constrained integers, then it shall
                    return a constrained integer whose constraint is calculated
                    by applying the operation to all possible value pairs. *)
-                let () =
+                let+ () =
                   match op with
                   | DIV ->
                       (* TODO cs1 divides cs1 ? How is it expressable in term of constraints? *)
-                      check_true' (constraints_is_strict_positive cs2) ()
+                      check_true' (constraints_is_strict_positive cs2)
                   | DIVRM | MOD ->
                       (* assert cs2 strict-positive *)
-                      check_true' (constraints_is_strict_positive cs2) ()
+                      check_true' (constraints_is_strict_positive cs2)
                   | SHL | SHR ->
                       (* assert cs2 non-negative *)
-                      check_true' (constraints_is_non_negative cs2) ()
-                  | _ -> ()
+                      check_true' (constraints_is_non_negative cs2)
+                  | _ -> fun () -> ()
+                in
+                let cs =
+                  best_effort UnConstrained (fun _ ->
```

Comment:
If I understand correctly, UnConstrained here is just a dummy value. If so, it is confusing since upon a first read it looks like you're creating an unconstrained integer from a bunch of constraints. Can you perhaps replace it with unit to make it less confusing?